src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,125 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROW_INDEX = 1.5;
+    private static final int START_INDEX_VALUE = 0;
+    private T[] dataArray;
+    private int size;
+
+    public ArrayList() {
+        dataArray = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int capacity) {
+        dataArray = (T[]) new Object[capacity];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == dataArray.length) {
+            dataArray = grow();
+        }
+        dataArray[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("index is invalid");
+        }
+        if (size == dataArray.length) {
+            dataArray = grow();
+        }
+        if (index == size) {
+            dataArray[size++] = value;
+        } else {
+            System.arraycopy(dataArray, index, dataArray, index + 1, size++);
+            dataArray[index] = value;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int oldSize = size;
+        int neededSize = size + list.size();
+        int indexOfValueFromList = START_INDEX_VALUE;
+        if (size == dataArray.length || dataArray.length < neededSize) {
+            dataArray = growToNeededSize(neededSize);
+        }
+        for (int i = oldSize; i < neededSize; i++) {
+            dataArray[i] = list.get(indexOfValueFromList++);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) dataArray[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        dataArray[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return remove(dataArray[index]);
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T[] temp = (T[]) new Object[Math.max(DEFAULT_CAPACITY, dataArray.length - 1)];
```

Comment:
Creating a new array `temp` with a reduced size every time an element is removed is inefficient. Instead, use `System.arraycopy` to shift the remaining elements to the left and nullify the last element to prevent memory leaks.